core/src/main/scala/kafka/server/ReplicaManager.scala

```diff
@@ -538,6 +539,92 @@ class ReplicaManager(val config: KafkaConfig,
     }
   }
 
+  /*
+   * For each pair of partition and log directory specified in the map, record the pair in the memory so that the partition
+   * will be created in the specified log directory when broker receives LeaderAndIsrRequest for the partition later.
+   *
+   * This API is currently only useful if the replica has not been created yet. We will be able to move replicas
+   * that are already created to the user-specified log directory after KIP-113 is fully implemented
+   *
+   */
+  def alterReplicaDir(partitionDirs: Map[TopicPartition, String]): Map[TopicPartition, Errors] = {
+    partitionDirs.map { case (topicPartition, destinationDir) =>
+      try {
+        if (!logManager.isLogDirOnline(destinationDir))
+          throw new KafkaStorageException(s"Log directory $destinationDir is offline")
+
+        // If the log for this partition has not been created yet:
+        // 1) Respond with ReplicaNotAvailableException for this partition in the AlterReplicaDirResponse
+        // 2) Record the destination log directory in the memory so that the partition will be created in this log directory
+        //    when broker receives LeaderAndIsrRequest for this partition later.
+        getReplica(topicPartition) match {
+          case Some(_) => // The support for moving replica between log directories on the same broker is not available yet.
+          case None =>
+            logManager.updatePreferredLogDir(topicPartition, destinationDir)
+            throw new ReplicaNotAvailableException(s"Replica $localBrokerId is not available for partition $topicPartition")
+        }
+
+        (topicPartition, Errors.NONE)
+      } catch {
+        case e@(_: LogDirNotFoundException |
+                _: ReplicaNotAvailableException |
+                _: KafkaStorageException) =>
+          (topicPartition, Errors.forException(e))
+        case t: Throwable =>
+          error("Error while changing replica dir for partition %s".format(topicPartition), t)
+          (topicPartition, Errors.forException(t))
+      }
+    }
+  }
+
+  /*
+   * Get the LogDirInfo for the specified list of log directories and partitions.
+   * All partitions will be queried if an empty set of partitions is specified.
+   *
+   * Each LogDirInfo specifies the following information for a given log directory:
+   * 1) Error of the log directory, e.g. whether the log is online or offline
+   * 2) size and lag of primary and temporary logs in the given log directory. A replica is included only if
+   *    its partition is queried. There may be temporary logs on the broker after KIP-113 is implemented.
+   */
+  def describeLogDirs(partitions: Set[TopicPartition]): Map[String, LogDirInfo] = {
```

Comment:
I don't think we should do that. Including this information in `DescribeLogDirRequest` will complicates the design of related APIs and expose an internal optimization detail to user. On the other hand, I don't think it provides any benefit to the user. The purpose of DescribeLogDirRequest is to get the load distribution of log directories on the broker so that user can determine a good reassignment replicas across log directories. Is there any information in `preferredLogDirs` that can help with this purpose of the `DescribeLogDirRequest`?